File: scripts/benchmark.py
import time
from ytapis import search, get_video
from ytapis.core import search_trending


def fmt(n: float) -> str:
    return f"{n:.1f}"


def pad(s: str, length: int) -> str:
    return s.ljust(length)


def render_table(rows: list[dict]) -> None:
    cols = ["Function", "Limit", "Time (ms)", "Results"]
    widths = [len(c) for c in cols]
    for row in rows:
        for i, c in enumerate(cols):
            widths[i] = max(widths[i], len(str(row[c])))
    sep = "+" + "+".join("-" * (w + 2) for w in widths) + "+"
    print(sep)
    header = "| " + " | ".join(pad(c, widths[i]) for i, c in enumerate(cols)) + " |"
    print(header)
    print(sep)
    for row in rows:
        line = "| " + " | ".join(pad(str(row[c]), widths[i]) for i, c in enumerate(cols)) + " |"
        print(line)
    print(sep)


def main() -> None:
    rows: list[dict] = []

    for limit in [5, 10, 20, 50]:
        try:
            t0 = time.perf_counter()
            results = search("cats", limit=limit)
            elapsed = (time.perf_counter() - t0) * 1000
            rows.append({"Function": "search", "Limit": str(limit), "Time (ms)": fmt(elapsed), "Results": str(len(results))})
        except Exception as e:
            rows.append({"Function": "search", "Limit": str(limit), "Time (ms)": "ERR", "Results": str(e)[:50]})

    for limit in [10]:
        try:
            t0 = time.perf_counter()
            results = search_trending(limit=limit)
            elapsed = (time.perf_counter() - t0) * 1000
            rows.append({"Function": "search_trending", "Limit": str(limit), "Time (ms)": fmt(elapsed), "Results": str(len(results))})
        except Exception as e:
            rows.append({"Function": "search_trending", "Limit": str(limit), "Time (ms)": "ERR", "Results": str(e)[:50]})

    try:
        t0 = time.perf_counter()
        video = get_video("dQw4w9WgXcQ")
        elapsed = (time.perf_counter() - t0) * 1000
        rows.append({"Function": "get_video", "Limit": "dQw4w9WgXcQ", "Time (ms)": fmt(elapsed), "Results": video.title})
    except Exception as e:
        rows.append({"Function": "get_video", "Limit": "dQw4w9WgXcQ", "Time (ms)": "ERR", "Results": str(e)[:50]})

    render_table(rows)


if __name__ == "__main__":
    main()

File: scripts/benchmark.ts
import { search, getVideo, searchTrending } from 'ytapis-core';

interface BenchmarkRow {
  Function: string;
  Limit: string;
  'Time (ms)': string;
  Results: string;
}

function fmt(n: number): string {
  return n.toFixed(1);
}

function pad(s: string, len: number): string {
  return s.padEnd(len);
}

function table(rows: BenchmarkRow[]) {
  const cols = ['Function', 'Limit', 'Time (ms)', 'Results'] as const;
  const widths = cols.map(c => Math.max(c.length, ...rows.map(r => String(r[c]).length)));
  const sep = '+' + widths.map(w => '-'.repeat(w + 2)).join('+') + '+';

  console.log(sep);
  const header = '| ' + cols.map((c, i) => pad(c, widths[i])).join(' | ') + ' |';
  console.log(header);
  console.log(sep);
  for (const row of rows) {
    const line = '| ' + cols.map((c, i) => pad(String(row[c]), widths[i])).join(' | ') + ' |';
    console.log(line);
  }
  console.log(sep);
}

async function time<T>(fn: () => Promise<T>): Promise<[number, T]> {
  const start = performance.now();
  const result = await fn();
  const elapsed = performance.now() - start;
  return [elapsed, result];
}

async function main() {
  const rows: BenchmarkRow[] = [];
  const limits = [5, 10, 20, 50];

  for (const limit of limits) {
    try {
      const [elapsed, res] = await time(() => search('cats', { limit }));
      rows.push({ Function: 'search', Limit: String(limit), 'Time (ms)': fmt(elapsed), Results: String(res.results.length) });
    } catch (e: any) {
      rows.push({ Function: 'search', Limit: String(limit), 'Time (ms)': 'ERR', Results: e.message });
    }
  }

  const limitsTrending = [10];
  for (const limit of limitsTrending) {
    try {
      const [elapsed, res] = await time(() => searchTrending({ limit }));
      rows.push({ Function: 'searchTrending', Limit: String(limit), 'Time (ms)': fmt(elapsed), Results: String(res.results.length) });
    } catch (e: any) {
      rows.push({ Function: 'searchTrending', Limit: String(limit), 'Time (ms)': 'ERR', Results: e.message });
    }
  }

  try {
    const [elapsed, res] = await time(() => getVideo('dQw4w9WgXcQ'));
    rows.push({ Function: 'getVideo', Limit: 'dQw4w9WgXcQ', 'Time (ms)': fmt(elapsed), Results: res.title });
  } catch (e: any) {
    rows.push({ Function: 'getVideo', Limit: 'dQw4w9WgXcQ', 'Time (ms)': 'ERR', Results: e.message });
  }

  table(rows);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
